src/main/java/net/snowflake/ingest/streaming/internal/ParquetValueParser.java

```diff
@@ -0,0 +1,334 @@
+                .getTimeInScale()
+                .longValue();
+        break;
+      case TIMESTAMP_TZ:
+        longValue =
+            DataValidationUtil.validateAndParseTimestampTz(value, scale)
+                .getSfTimestamp()
+                .orElseThrow(
+                    () ->
+                        new SFException(
+                            ErrorCode.INVALID_ROW,
+                            value,
+                            "Unable to parse timestamp for TIMESTAMP_TZ column"))
+                .toBinary(scale, true)
+                .longValue();
+        break;
+      case FIXED:
+        BigDecimal bigDecimalValue = DataValidationUtil.validateAndParseBigDecimal(value);
+        bigDecimalValue = bigDecimalValue.setScale(scale, RoundingMode.HALF_UP);
+        checkValueInRange(bigDecimalValue, scale, precision);
+        longValue = bigDecimalValue.longValue();
+        break;
+      default:
+        throw new SFException(ErrorCode.UNKNOWN_DATA_TYPE, logicalType, physicalType);
+    }
+    return longValue;
+  }
+
+  /**
+   * Parses an int128 value based on Snowflake logical type.
+   *
+   * @param value column value provided by user in a row
+   * @param scale data type scale
+   * @param precision data type precision
+   * @param logicalType Snowflake logical type
+   * @param physicalType Snowflake physical type
+   * @return parsed int64 value
+   */
+  private static BigInteger getSb16Value(
+      Object value,
+      int scale,
+      int precision,
+      AbstractRowBuffer.ColumnLogicalType logicalType,
+      AbstractRowBuffer.ColumnPhysicalType physicalType) {
+    switch (logicalType) {
+      case TIMESTAMP_TZ:
+        return DataValidationUtil.validateAndParseTimestampTz(value, scale)
+            .getSfTimestamp()
+            .orElseThrow(
+                () ->
+                    new SFException(
+                        ErrorCode.INVALID_ROW,
+                        value,
+                        "Unable to parse timestamp for TIMESTAMP_TZ column"))
+            .toBinary(scale, true);
+      case TIMESTAMP_LTZ:
+      case TIMESTAMP_NTZ:
+        boolean ignoreTimezone = logicalType == AbstractRowBuffer.ColumnLogicalType.TIMESTAMP_NTZ;
+        return DataValidationUtil.validateAndParseTimestampNtzSb16(value, scale, ignoreTimezone)
+            .getTimeInScale();
+      case FIXED:
+        BigDecimal bigDecimalValue = DataValidationUtil.validateAndParseBigDecimal(value);
+        // explicitly match the BigDecimal input scale with the Snowflake data type scale
+        bigDecimalValue = bigDecimalValue.setScale(scale, RoundingMode.HALF_UP);
+        checkValueInRange(bigDecimalValue, scale, precision);
+        return bigDecimalValue.unscaledValue();
+      default:
+        throw new SFException(ErrorCode.UNKNOWN_DATA_TYPE, logicalType, physicalType);
+    }
+  }
+
+  private static void checkValueInRange(BigDecimal bigDecimalValue, int scale, int precision) {
```

Comment:
I believe we have similar logic for Arrow as well, could you combine them?

Reply:
ack.